src/litdata/utilities/dataset_utilities.py

```diff
@@ -0,0 +1,113 @@
+
+    # Make sure input_dir contains cache path and remote url
+    if _should_replace_path(input_dir.path):
+        cache_path = _try_create_cache_dir(input_dir=input_dir.path if input_dir.path else input_dir.url)
+        if cache_path is not None:
+            input_dir.path = cache_path
+
+    assert input_dir.path is not None
+
+    cache_index_filepath = os.path.join(input_dir.path, _INDEX_FILENAME)
+
+    # Check if `index.json` file exists in cache path
+    if not os.path.exists(cache_index_filepath) and isinstance(input_dir.url, str):
+        assert input_dir.url is not None
+        downloader = get_downloader_cls(input_dir.url, input_dir.path, [])
+        downloader.download_file(os.path.join(input_dir.url, _INDEX_FILENAME), cache_index_filepath)
+
+    if os.path.exists(os.path.join(input_dir.path, _INDEX_FILENAME)):
+        # load chunks from `index.json` file
+        with open(os.path.join(input_dir.path, _INDEX_FILENAME)) as f:
+            data = json.load(f)
+            original_chunks = data["chunks"]
+    else:
+        raise ValueError(
+            f"The provided dataset `{input_dir.path}` doesn't contain any {_INDEX_FILENAME} file."
+            " HINT: Did you successfully optimize a dataset to the provided `input_dir`?"
+        )
+
+    assert len(original_chunks) > 0, f"No chunks found in the `{input_dir}/index.json` file"
+
+    # create a (chunk_start, chunk_end) list to indicate our subsample from where we can read.
+    my_roi = generate_roi(original_chunks, item_loader)
+
+    # shuffle lists together
+    if shuffle and not math.isclose(subsample, 1.0):
+        # checking if subsample is 1, as if user wants complete data, then let, shuffler and sampler do the work
+        original_chunks, my_roi = shuffle_lists_together(original_chunks, my_roi, seed)
+
+    target = int(sum([roi[1] - roi[0] for roi in my_roi]) * subsample)
+
+    my_subsampled_files, my_roi, _, _ = my_subsampled_filenames_and_roi(original_chunks, my_roi, target)
+
+    return my_subsampled_files, my_roi
+
+
+def _should_replace_path(path: Optional[str]) -> bool:
+    """Whether the input path is a special path to be replaced."""
+    if path is None or path == "":
+        return True
+
+    return path.startswith("/teamspace/datasets/") or path.startswith("/teamspace/s3_connections/")
+
+
+def _try_create_cache_dir(input_dir: Optional[str]) -> Optional[str]:
+    hash_object = hashlib.md5((input_dir or "").encode())  # noqa: S324
+    if "LIGHTNING_CLUSTER_ID" not in os.environ or "LIGHTNING_CLOUD_PROJECT_ID" not in os.environ:
+        cache_dir = os.path.join(_DEFAULT_CACHE_DIR, hash_object.hexdigest())
+        os.makedirs(cache_dir, exist_ok=True)
+        return cache_dir
+    cache_dir = os.path.join("/cache", "chunks", hash_object.hexdigest())
+    os.makedirs(cache_dir, exist_ok=True)
+    return cache_dir
+
+
+def generate_roi(chunks: List[Dict[str, Any]], item_loader: Optional[BaseItemLoader] = None) -> List[Tuple[int, int]]:
```

Comment:
This is `item_loader.generate_intervals()`. No need to re-implement it. This can lead to issues in the future if they get out of sync.

Reply:
But, Both are different. Original one create intervals like: `[(0,50), (50, 100), (100,150)..]` and this one creates `[(0,50), (0, 50), (0, 50)..]`.

It is so bcoz, no matter how you shuffle, but for this chunk, you can read from (X, Y).

When item_loader actually generates the interval, here's the code:

```python
def generate_intervals(self) -> List[Tuple[int, int, int, int]]:
        intervals = []
        begin = 0
        end = 0
        for idx, curr_chunk in enumerate(self._chunks):
            end += curr_chunk["chunk_size"]
           # -------------------------------------------------------------------
            start_idx, end_idx = begin, end # it will be overlap in most cases
            if self.region_of_interest is not None:
                start_idx = begin + self.region_of_interest[idx][0] # when you are allowed to ready from a subset of chunk
                end_idx = begin + self.region_of_interest[idx][1] # when you are allowed to ready from a subset of chunk
           # -------------------------------------------------------------------
            intervals.append((begin, start_idx, end_idx, end))
            begin += curr_chunk["chunk_size"]
        return intervals
```

Reply:
Good point.